src/pextlib1.0/Pextlib.c

```diff
@@ -649,10 +651,149 @@ int SetMaxOpenFilesCmd(ClientData clientData UNUSED, Tcl_Interp *interp, int obj
 	return TCL_OK;
 }
 
+/* Mount point file system case-sensitivity caching infrastructure. */
+typedef struct _mount_cs_cache_entry {
+    char *mountpoint;
+    int case_sensitive;
+} mount_cs_cache_entry_t, *mount_cs_cache_entry_list_t;
+
+struct _mount_cs_cache {
+    size_t count;
+    mount_cs_cache_entry_t **entries;
+};
+
+/**
+ * Returns a new pre-allocated mount_cs_cache_t object.
+ */
+mount_cs_cache_t* new_mount_cs_cache() {
+    mount_cs_cache_t *ret = malloc(sizeof(mount_cs_cache_t));
+
+    if (ret) {
+        ret->count = 0;
+        ret->entries = NULL;
+    }
+
+    return ret;
+}
+
+/**
+ * Resets a mount cache object.
+ */
+void reset_mount_cs_cache(mount_cs_cache_t *cache) {
+    if (cache) {
+        /*
+         * Assume that if the count of cached entries is zero,
+         * the entries handle is set to NULL as well. Likewise,
+         * if the count is non-zero, the entries pointer should
+         * be non-NULL. Any other combination probably means
+         * something is very, very wrong.
+         */
+        assert((!cache->count && !cache->entries) || (cache->count && cache->entries));
+
+        for (size_t i = 0; i < cache->count; ++i) {
+            free(cache->entries[i]->mountpoint);
+            cache->entries[i]->mountpoint = NULL;
+
+            free(cache->entries[i]);
+            cache->entries[i] = NULL;
+        }
+
+        free(cache->entries);
+        cache->entries = NULL;
+    }
+}
+
+/**
+ * Adds a new entry to a mount cache object.
+ * Returns zero on success, -1 on failure.
+ */
+static int add_to_mount_cs_cache(Tcl_Interp *interp, mount_cs_cache_t *cache, const char *mountpoint, int case_sensitive) {
+    int ret = -1;
+
+    if ((cache) && (mountpoint)) {
+        /*
+         * Go the greedy road and resize as needed. Good enough probably, since
+         * systems normally don't have a huge amount of mounts to start with.
+         */
+        mount_cs_cache_entry_t **new_data = realloc(cache->entries, (cache->count + 1) * sizeof(mount_cs_cache_entry_list_t));
+
+        if (!new_data) {
+            ui_info(interp, "pextlib: unable to reallocate FS cache entries list, leaving untouched.");
+            return ret;
+        }
+
+        cache->entries = new_data;
+
+        mount_cs_cache_entry_t *new_entry = malloc(sizeof(mount_cs_cache_entry_t));
+        new_entry->mountpoint = strdup(mountpoint);
```

Comment:
Strictly speaking, you should probably also have error handling for this allocation.